File: src/main/python/net/coderextreme/data/SimpleShader.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.0"
head1 = x3d.head()
head1.children.append(x3d.Comment("""meta content='under development' name='warning'/"""))
component2 = x3d.component()
component2.name = "Shaders"
component2.level = 1

head1.children.append(component2)
meta3 = x3d.meta()
meta3.name = "title"
meta3.content = "SimpleShader.x3d"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "description"
meta4.content = "Simple shader example"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "creator"
meta5.content = "John Stewart"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "translator"
meta6.content = "Holger Seelig, John Carlson, Don Brutzman and Jeff Malnick"

head1.children.append(meta6)
meta7 = x3d.meta()
meta7.name = "created"
meta7.content = "26 May 2009"

head1.children.append(meta7)
meta8 = x3d.meta()
meta8.name = "translated"
meta8.content = "15 October 2009"

head1.children.append(meta8)
meta9 = x3d.meta()
meta9.name = "modified"
meta9.content = "20 October 2019"

head1.children.append(meta9)
meta10 = x3d.meta()
meta10.name = "reference"
meta10.content = "http://www.opengl.org/wiki/Fragment_Shader"

head1.children.append(meta10)
meta11 = x3d.meta()
meta11.name = "reference"
meta11.content = "http://www.opengl.org/wiki/http://www.opengl.org/wiki/Vertex_Shader"

head1.children.append(meta11)
meta12 = x3d.meta()
meta12.name = "reference"
meta12.content = "https://www.web3d.org/x3d/wiki/index.php/X3D_Plugfest"

head1.children.append(meta12)
meta13 = x3d.meta()
meta13.name = "subject"
meta13.content = "X3D shader example"

head1.children.append(meta13)
meta14 = x3d.meta()
meta14.name = "reference"
meta14.content = "originals/simpleShader.x3dv"

head1.children.append(meta14)
meta15 = x3d.meta()
meta15.name = "reference"
meta15.content = "ShaderTutorialInstantReality.pdf"

head1.children.append(meta15)
meta16 = x3d.meta()
meta16.name = "generator"
meta16.content = "Titania V3.0.3, http://titania.create3000.de"

head1.children.append(meta16)
meta17 = x3d.meta()
meta17.name = "info"
meta17.content = "World of Titania"

head1.children.append(meta17)
meta18 = x3d.meta()
meta18.name = "outputStyle"
meta18.content = "nicest"

head1.children.append(meta18)
meta19 = x3d.meta()
meta19.name = "generator"
meta19.content = "X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"

head1.children.append(meta19)
meta20 = x3d.meta()
meta20.name = "identifier"
meta20.content = "https://www.web3d.org/x3d/content/examples/Basic/Shaders/SimpleShader.x3d"

head1.children.append(meta20)
meta21 = x3d.meta()
meta21.name = "license"
meta21.content = "../../license.html"

head1.children.append(meta21)

X3D0.head = head1
Scene22 = x3d.Scene()
ProtoDeclare23 = x3d.ProtoDeclare()
ProtoDeclare23.name = "myPrototype"
ProtoInterface24 = x3d.ProtoInterface()
field25 = x3d.field()
field25.name = "myInputRange"
field25.accessType = "initializeOnly"
field25.type = "SFVec3f"
field25.value = [0.95,0.44,0.22]

ProtoInterface24.field.append(field25)

ProtoDeclare23.ProtoInterface = ProtoInterface24
ProtoBody26 = x3d.ProtoBody()
Transform27 = x3d.Transform(DEF="TR")
Shape28 = x3d.Shape()
Appearance29 = x3d.Appearance()
Material30 = x3d.Material()
Material30.diffuseColor = [0.5,0.5,0.9]

Appearance29.material = Material30
ComposedShader31 = x3d.ComposedShader()
ComposedShader31.language = "GLSL"
field32 = x3d.field()
field32.name = "decis"
field32.accessType = "initializeOnly"
field32.type = "SFVec3f"

ComposedShader31.field.append(field32)
IS33 = x3d.IS()
connect34 = x3d.connect()
connect34.nodeField = "decis"
connect34.protoField = "myInputRange"

IS33.connect.append(connect34)

ComposedShader31.IS = IS33
ShaderPart35 = x3d.ShaderPart()
ShaderPart35.type = "VERTEX"

ComposedShader31.parts.append(ShaderPart35)
ShaderPart36 = x3d.ShaderPart()
ShaderPart36.type = "FRAGMENT"

ComposedShader31.parts.append(ShaderPart36)

Appearance29.shaders.append(ComposedShader31)
ComposedShader37 = x3d.ComposedShader(DEF="Cobweb")
ComposedShader37.language = "GLSL"
field38 = x3d.field()
field38.name = "decis"
field38.accessType = "initializeOnly"
field38.type = "SFVec3f"
field38.value = [0.95,0.77,0.44]

ComposedShader37.field.append(field38)
ShaderPart39 = x3d.ShaderPart()
ShaderPart39.type = "VERTEX"

ComposedShader37.parts.append(ShaderPart39)
ShaderPart40 = x3d.ShaderPart(DEF="_1")
ShaderPart40.type = "FRAGMENT"

ComposedShader37.parts.append(ShaderPart40)

Appearance29.shaders.append(ComposedShader37)

Shape28.appearance = Appearance29
Sphere41 = x3d.Sphere()
Sphere41.radius = 1.75

Shape28.geometry = Sphere41

Transform27.children.append(Shape28)

ProtoBody26.children.append(Transform27)

ProtoDeclare23.ProtoBody = ProtoBody26

Scene22.children.append(ProtoDeclare23)
WorldInfo42 = x3d.WorldInfo()
WorldInfo42.title = "SimpleShader"
MetadataSet43 = x3d.MetadataSet(DEF="Titania")
MetadataSet43.name = "Titania"
MetadataSet43.reference = "http://titania.create3000.de"
MetadataSet44 = x3d.MetadataSet(DEF="Selection")
MetadataSet44.name = "Selection"
MetadataSet44.reference = "http://titania.create3000.de"
MetadataSet45 = x3d.MetadataSet(DEF="nodes")
MetadataSet45.name = "nodes"
MetadataSet45.reference = "http://titania.create3000.de"
MetadataSet45.children.append(x3d.Comment("""NULL"""))

if MetadataSet44.value is None:
    MetadataSet44.value = []
MetadataSet44.value.append(MetadataSet45)

if MetadataSet43.value is None:
    MetadataSet43.value = []
MetadataSet43.value.append(MetadataSet44)
MetadataSet46 = x3d.MetadataSet(DEF="NavigationInfo")
MetadataSet46.name = "NavigationInfo"
MetadataSet46.reference = "http://titania.create3000.de"
MetadataString47 = x3d.MetadataString(DEF="type")
MetadataString47.name = "type"
MetadataString47.reference = "http://titania.create3000.de"
MetadataString47.value = ["EXAMINE"]

if MetadataSet46.value is None:
    MetadataSet46.value = []
MetadataSet46.value.append(MetadataString47)

if MetadataSet43.value is None:
    MetadataSet43.value = []
MetadataSet43.value.append(MetadataSet46)
MetadataSet48 = x3d.MetadataSet(DEF="Viewpoint")
MetadataSet48.name = "Viewpoint"
MetadataSet48.reference = "http://titania.create3000.de"
MetadataDouble49 = x3d.MetadataDouble(DEF="position")
MetadataDouble49.name = "position"
MetadataDouble49.reference = "http://titania.create3000.de"
MetadataDouble49.value = [6.24067728185014,0.00250837343276661,2.92117542307615]

if MetadataSet48.value is None:
    MetadataSet48.value = []
MetadataSet48.value.append(MetadataDouble49)
MetadataDouble50 = x3d.MetadataDouble(DEF="orientation")
MetadataDouble50.name = "orientation"
MetadataDouble50.reference = "http://titania.create3000.de"
MetadataDouble50.value = [-0.110173424710488,0.990158061907379,-0.0863065984000336,1.21146676119191]

if MetadataSet48.value is None:
    MetadataSet48.value = []
MetadataSet48.value.append(MetadataDouble50)
MetadataDouble51 = x3d.MetadataDouble(DEF="centerOfRotation")
MetadataDouble51.name = "centerOfRotation"
MetadataDouble51.reference = "http://titania.create3000.de"
MetadataDouble51.value = [-0.808320198626341,-0.358072370409949,0.22817191560906]

if MetadataSet48.value is None:
    MetadataSet48.value = []
MetadataSet48.value.append(MetadataDouble51)

if MetadataSet43.value is None:
    MetadataSet43.value = []
MetadataSet43.value.append(MetadataSet48)

WorldInfo42.metadata = MetadataSet43

Scene22.children.append(WorldInfo42)
ProtoInstance52 = x3d.ProtoInstance()
ProtoInstance52.name = "myPrototype"

Scene22.children.append(ProtoInstance52)

X3D0.Scene = Scene22
f = open("../data/SimpleShader.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/SimpleShader.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/SimpleShader.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/go/net/coderextreme/data/SimpleShader.go

package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"gox3d/x3d" // Assumes your module is named x3d-go-tester

	// Imports for the CGo-based libxml2 wrapper
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

// ... (Helper functions remain the same) ...
func stringPtr(s string) *string { return &s }
func floatPtr(f float32) *float32 { return &f }
func doublePtr(d float64) *float64 { return &d }
func boolPtr(b bool) *bool       { return &b }
func int32Ptr(i int32) *int32    { return &i }

func downloadSchemaIfNotExists(url, filepath string) error {
	// ... (function is correct, no changes) ...
	if _, err := os.Stat(filepath); err == nil {
		fmt.Printf("Schema file '%s' already exists. Skipping download.\n", filepath)
		return nil
	}
	fmt.Printf("Downloading X3D schema from %s to %s...\n", url, filepath)
	resp, err := http.Get(url)
	if err != nil { return fmt.Errorf("failed to download schema: %w", err) }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return fmt.Errorf("bad status: %s", resp.Status) }
	out, err := os.Create(filepath)
	if err != nil { return fmt.Errorf("failed to create schema file: %w", err) }
	defer out.Close()
	_, err = io.Copy(out, resp.Body)
	if err != nil { return fmt.Errorf("failed to save schema: %w", err) }
	fmt.Println("Schema downloaded successfully.")
	return nil
}

// Updated validation function to print detailed errors
func validateXMLWithSchema(xmlData []byte, schemaPath string) error {
	schema, err := xsd.ParseFromFile(schemaPath)
	if err != nil { return fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err) }
	defer schema.Free()
	doc, err := libxml2.Parse(xmlData)
	if err != nil { return fmt.Errorf("failed to parse XML data: %w", err) }
	defer doc.Free()

	validationErr := schema.Validate(doc)
	if validationErr != nil {
		errorLog := ""
		if schemaErrors, ok := validationErr.(xsd.SchemaValidationError); ok {
			for i, e := range schemaErrors.Errors() {
				errorLog += fmt.Sprintf("Error %d: %s\n", i+1, e.Error())
			}
			return fmt.Errorf("schema validation failed with %d errors:\n%s", len(schemaErrors.Errors()), errorLog)
		}
		return fmt.Errorf("schema validation failed: %w", validationErr)
	}
	return nil
}

func main() {
	fmt.Println("--- Building and Testing an X3D Scene in Go ---")

	const schemaURL = "https://www.web3d.org/specifications/x3d-4.1.xsd"
	const schemaFilename = "x3d-4.1.xsd"
	if err := downloadSchemaIfNotExists(schemaURL, schemaFilename); err != nil {
		log.Fatalf("Could not prepare schema file: %v", err)
	}
    sceneRoot := &x3d.X3D{
        Profile: stringPtr("Immersive"),
        Version: stringPtr("4.0"),
        Head: &x3d.Head{
//meta content='under development' name='warning'/
            Components: []*x3d.Component{
                &x3d.Component{
                    Name: stringPtr("Shaders"),
                    Level: int32Ptr(1),
            },
        },
            Metas: []*x3d.Meta{
                &x3d.Meta{
                    Name: stringPtr("title"),
                    Content: stringPtr("SimpleShader.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("description"),
                Content: stringPtr("Simple shader example"),
            },
            &x3d.Meta{
                Name: stringPtr("creator"),
                Content: stringPtr("John Stewart"),
            },
            &x3d.Meta{
                Name: stringPtr("translator"),
                Content: stringPtr("Holger Seelig, John Carlson, Don Brutzman and Jeff Malnick"),
            },
            &x3d.Meta{
                Name: stringPtr("created"),
                Content: stringPtr("26 May 2009"),
            },
            &x3d.Meta{
                Name: stringPtr("translated"),
                Content: stringPtr("15 October 2009"),
            },
            &x3d.Meta{
                Name: stringPtr("modified"),
                Content: stringPtr("20 October 2019"),
            },
            &x3d.Meta{
                Name: stringPtr("reference"),
                Content: stringPtr("http://www.opengl.org/wiki/Fragment_Shader"),
            },
            &x3d.Meta{
                Name: stringPtr("reference"),
                Content: stringPtr("http://www.opengl.org/wiki/http://www.opengl.org/wiki/Vertex_Shader"),
            },
            &x3d.Meta{
                Name: stringPtr("reference"),
                Content: stringPtr("https://www.web3d.org/x3d/wiki/index.php/X3D_Plugfest"),
            },
            &x3d.Meta{
                Name: stringPtr("subject"),
                Content: stringPtr("X3D shader example"),
            },
            &x3d.Meta{
                Name: stringPtr("reference"),
                Content: stringPtr("originals/simpleShader.x3dv"),
            },
            &x3d.Meta{
                Name: stringPtr("reference"),
                Content: stringPtr("ShaderTutorialInstantReality.pdf"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("Titania V3.0.3, http://titania.create3000.de"),
            },
            &x3d.Meta{
                Name: stringPtr("info"),
                Content: stringPtr("World of Titania"),
            },
            &x3d.Meta{
                Name: stringPtr("outputStyle"),
                Content: stringPtr("nicest"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"),
            },
            &x3d.Meta{
                Name: stringPtr("identifier"),
                Content: stringPtr("https://www.web3d.org/x3d/content/examples/Basic/Shaders/SimpleShader.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("license"),
                Content: stringPtr("../../license.html"),
            },
            },
        },
        &x3d.Group{
            Children: []x3d.X3DNode{
                &x3d.ProtoDeclare{
                    Name: stringPtr("myPrototype"),
                    ProtoInterface: &x3d.ProtoInterface{
                        Field: []x3d.X3DNode{
                            &x3d.Field{
                                Name: stringPtr("myInputRange"),
                                AccessType: stringPtr("initializeOnly"),
                                Type: stringPtr("SFVec3f"),
                                Value: stringPtr("0.95 0.44 0.22"),
                        },
                        },
                    },
                    ProtoBody: &x3d.ProtoBody{
                        Children: []x3d.X3DNode{
                            &x3d.Transform{
                                    CoreX3DNode: x3d.CoreX3DNode{
                                        DEF: stringPtr("TR"),
                                    },
                                Children: []x3d.X3DNode{
                                    &x3d.Shape{
                                        Appearance: &x3d.Appearance{
                                            Material: &x3d.Material{
                                                DiffuseColor: &x3d.SFColor{0.5, 0.5, 0.9},
                                            },
                                            &x3d.ComposedShader{
                                                Language: stringPtr("GLSL"),
                                                IS: &x3d.IS{
                                                    Connect: []x3d.X3DNode{
                                                        &x3d.Connect{
                                                            NodeField: stringPtr("decis"),
                                                            ProtoField: stringPtr("myInputRange"),
                                                    },
                                                    },
                                                },
                                                Field: []x3d.X3DNode{
                                                    &x3d.Field{
                                                        Name: stringPtr("decis"),
                                                        AccessType: stringPtr("initializeOnly"),
                                                        Type: stringPtr("SFVec3f"),
                                                },
                                            },
                                                Parts: []x3d.X3DNode{
                                                    &x3d.ShaderPart{
                                                        Type: stringPtr("VERTEX"),
                                                },
                                                &x3d.ShaderPart{
                                                    Type: stringPtr("FRAGMENT"),
                                                },
                                                },
                                            },
                                            &x3d.ComposedShader{
                                                CoreX3DNode: x3d.CoreX3DNode{
                                                    DEF: stringPtr("Cobweb"),
                                                },
                                                Language: stringPtr("GLSL"),
                                                Field: []x3d.X3DNode{
                                                    &x3d.Field{
                                                        Name: stringPtr("decis"),
                                                        AccessType: stringPtr("initializeOnly"),
                                                        Type: stringPtr("SFVec3f"),
                                                        Value: stringPtr("0.95 0.77 0.44"),
                                                },
                                            },
                                                Parts: []x3d.X3DNode{
                                                    &x3d.ShaderPart{
                                                        Type: stringPtr("VERTEX"),
                                                },
                                                &x3d.ShaderPart{
                                                    CoreX3DNode: x3d.CoreX3DNode{
                                                        DEF: stringPtr("_1"),
                                                    },
                                                    Type: stringPtr("FRAGMENT"),
                                                },
                                                },
                                            },
                                        },
                                        Geometry: &x3d.Sphere{
                                            Radius: floatPtr(1.75),
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.WorldInfo{
                    Title: stringPtr("SimpleShader"),
                    Metadata: &x3d.MetadataSet{
                        Name: stringPtr("Titania"),
                        CoreX3DNode: x3d.CoreX3DNode{
                            DEF: stringPtr("Titania"),
                        },
                        Reference: stringPtr("http://titania.create3000.de"),
                        &x3d.MetadataSet{
                            Name: stringPtr("Selection"),
                            CoreX3DNode: x3d.CoreX3DNode{
                                DEF: stringPtr("Selection"),
                            },
                            Reference: stringPtr("http://titania.create3000.de"),
                            &x3d.MetadataSet{
                                Name: stringPtr("nodes"),
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("nodes"),
                                },
                                Reference: stringPtr("http://titania.create3000.de"),
//NULL
                            },
                        },
                        &x3d.MetadataSet{
                            Name: stringPtr("NavigationInfo"),
                            CoreX3DNode: x3d.CoreX3DNode{
                                DEF: stringPtr("NavigationInfo"),
                            },
                            Reference: stringPtr("http://titania.create3000.de"),
                            &x3d.MetadataString{
                                Name: stringPtr("type"),
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("type"),
                                },
                                Reference: stringPtr("http://titania.create3000.de"),
                                Value: x3d.MFString{"EXAMINE"},
                            },
                        },
                        &x3d.MetadataSet{
                            Name: stringPtr("Viewpoint"),
                            CoreX3DNode: x3d.CoreX3DNode{
                                DEF: stringPtr("Viewpoint"),
                            },
                            Reference: stringPtr("http://titania.create3000.de"),
                            &x3d.MetadataDouble{
                                Name: stringPtr("position"),
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("position"),
                                },
                                Reference: stringPtr("http://titania.create3000.de"),
                                Value: x3d.MFDouble{6.24067728185014, 0.00250837343276661, 2.92117542307615},
                            },
                            &x3d.MetadataDouble{
                                Name: stringPtr("orientation"),
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("orientation"),
                                },
                                Reference: stringPtr("http://titania.create3000.de"),
                                Value: x3d.MFDouble{-0.110173424710488, 0.990158061907379, -0.0863065984000336, 1.21146676119191},
                            },
                            &x3d.MetadataDouble{
                                Name: stringPtr("centerOfRotation"),
                                CoreX3DNode: x3d.CoreX3DNode{
                                    DEF: stringPtr("centerOfRotation"),
                                },
                                Reference: stringPtr("http://titania.create3000.de"),
                                Value: x3d.MFDouble{-0.808320198626341, -0.358072370409949, 0.22817191560906},
                            },
                        },
                    },
                },
                &x3d.ProtoInstance{
                    Name: stringPtr("myPrototype"),
                },
            },
        },
    }


	fmt.Println("\n--- Validating the generated scene (internal logic) ---")
	err := sceneRoot.Scene.Validate()
	if err != nil {
		log.Fatalf("Scene validation failed: %v", err)
	}
	fmt.Println("✅ Internal scene validation successful!")
	fmt.Println("\n--- Serializing scene to XML ---")
	output, err := xml.MarshalIndent(sceneRoot, "", "  ")

	if err != nil {
		log.Fatalf("XML Marshaling failed: %v", err)
	}
	/*
	fmt.Println("\n--- Validating XML against X3D 4.1 Schema (using libxml2) ---")
	err = validateXMLWithSchema(output, schemaFilename)
	if err != nil {
		fmt.Printf("--- Invalid Generated XML ---\n%s\n---------------------------\n", string(output))
		log.Fatalf("Schema validation failed for generated XML: %v", err)
	}
	fmt.Println("✅ XML is valid against the X3D 4.1 schema!")
	*/
	filename := "../data/SimpleShader.new.go.x3d"
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // Ensure the file is closed when the function exits

	// Write the string content to the file
	header := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE X3D PUBLIC \"ISO//Web3D//DTD X3D 4.1//EN\" \"https://www.web3d.org/specifications/x3d-4.1.dtd\">\n"
	_, err = file.WriteString(header)
	if err != nil {
		fmt.Printf("Error writing header to file: %v\n", err)
		return
	}

	_, err = file.Write(output)
	if err != nil {
		fmt.Printf("Error writing bytes to file: %v\n", err)
		return
	}

	fmt.Printf("\n✅ Scene successfully written to %s\n", filename)
	fmt.Printf("   You can view the generated XML below:\n\n%s\n", string(output))
}